realm/realm-library/src/main/cpp/io_realm_internal_OsRealmConfig.cpp

```diff
@@ -0,0 +1,332 @@
+
+#include <shared_realm.hpp>
+#if REALM_ENABLE_SYNC
+#include <sync/sync_config.hpp>
+#include <sync/sync_manager.hpp>
+#include <sync/sync_session.hpp>
+#endif
+
+#include "util.hpp"
+#include "jni_util/java_method.hpp"
+#include "jni_util/java_class.hpp"
+#include "jni_util/java_global_ref.hpp"
+#include "jni_util/jni_utils.hpp"
+
+using namespace realm;
+using namespace realm::jni_util;
+
+static_assert(SchemaMode::Automatic ==
+                  static_cast<SchemaMode>(io_realm_internal_OsRealmConfig_SCHEMA_MODE_VALUE_AUTOMATIC),
+              "");
+static_assert(SchemaMode::ReadOnly ==
+                  static_cast<SchemaMode>(io_realm_internal_OsRealmConfig_SCHEMA_MODE_VALUE_READONLY),
+              "");
+static_assert(SchemaMode::ResetFile ==
+                  static_cast<SchemaMode>(io_realm_internal_OsRealmConfig_SCHEMA_MODE_VALUE_RESET_FILE),
+              "");
+static_assert(SchemaMode::Additive ==
+                  static_cast<SchemaMode>(io_realm_internal_OsRealmConfig_SCHEMA_MODE_VALUE_ADDITIVE),
+              "");
+static_assert(SchemaMode::Manual == static_cast<SchemaMode>(io_realm_internal_OsRealmConfig_SCHEMA_MODE_VALUE_MANUAL),
+              "");
+
+static void finalize_realm_config(jlong ptr)
+{
+    TR_ENTER_PTR(ptr)
+    delete reinterpret_cast<Realm::Config*>(ptr);
+}
+
+static JavaClass& get_shared_realm_class(JNIEnv* env)
+{
+    static JavaClass shared_realm_class(env, "io/realm/internal/SharedRealm");
+    return shared_realm_class;
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_OsRealmConfig_nativeGetFinalizerPtr(JNIEnv*, jclass)
+{
+    TR_ENTER()
+    return reinterpret_cast<jlong>(&finalize_realm_config);
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_OsRealmConfig_nativeCreate(JNIEnv* env, jclass, jstring j_realm_path,
+                                                                          jboolean enable_cache,
+                                                                          jboolean enable_format_upgrade)
+{
+    TR_ENTER()
+    try {
+        JStringAccessor realm_path(env, j_realm_path);
+        auto* config_ptr = new Realm::Config();
+        config_ptr->path = realm_path;
+        config_ptr->cache = enable_cache;
+        config_ptr->disable_format_upgrade = !enable_format_upgrade;
+        return reinterpret_cast<jlong>(config_ptr);
+    }
+    CATCH_STD()
+    return reinterpret_cast<jlong>(nullptr);
+}
+
+JNIEXPORT void JNICALL Java_io_realm_internal_OsRealmConfig_nativeSetEncryptionKey(JNIEnv* env, jclass,
+                                                                                   jlong native_ptr,
+                                                                                   jbyteArray j_key_array)
+{
+    TR_ENTER_PTR(native_ptr)
+    try {
+        JniByteArray key_array(env, j_key_array);
+        auto& config = *reinterpret_cast<Realm::Config*>(native_ptr);
+        // Encryption key should be set before creating sync_config.
```

Comment:
Because of we need to get the key from `Realm::config` and set it to `SyncConfig`. https://github.com/realm/realm-java/pull/5065/files/e9f892215966160978abbe428ad24de5e23ce9b9#diff-99d5c7b90ca048df8fa6511c6daec6c7R291

This is actually a very strange API design in Object Store :(